File: Recursion and Backtracking/NQueens.java
import java.util.Scanner;

class NQueens {

  static boolean Solve(int[][] board, int col) {
    // This base case means that, we are done with the last row of placing the Queen and there is no
    // next column left.
    if (col >= board.length) return true;
    // Trying for each row
    for (int rowToTry = 0; rowToTry < board.length; rowToTry++) {
      if (IsSafe(board, rowToTry, col)) {
        PlaceQueen(board, rowToTry, col); // Place the queen if no attack is possible
        if (Solve(board, col + 1)) // Recurse For next Queen placement in next Column.
        return true; // if that choice works fine, then go for next Colum
        RemoveQueen(
            board, rowToTry,
            col); // else backtrack( or go to previous state, remove the placement and try again).
      }
    }
    return false;
  }

  // Place the queen if the position is safe
  static void PlaceQueen(int board[][], int row, int col) {
    board[row][col] = 1;
  }

  // Remove the Previous state of queen placement if decision seems to b wrong.
  static void RemoveQueen(int board[][], int row, int col) {
    board[row][col] = 0;
  }

  // Check if it is safe to place the queen at a particular location. If not returns false
  static boolean IsSafe(int[][] board, int row, int col) {
    //				System.out.println("Checking for point " + row + " "+ col);
    //				print(board, 4,4);

    //				System.out.println();

    // Row Check for conflict
    for (int i = 0; i < board.length; i++) {
      //		System.out.println("for index " +row+"," + i +"value "+ board[row][i]);
      if (board[row][i] == 1) {
        //								System.out.println("conflict at "+i+" "+col);
        return false;
      }
    }

    // Column Check for conflict
    for (int i = 0; i < board[row].length; i++) {
      if (board[i][col] == 1) return false;
    }

    // Diagonal Right-Up check for conflict
    int i = row;
    int j = col;
    while (i < board.length && j < board[i].length) {
      if (board[i][j] == 1) return false;
      i++;
      j++;
    }

    // Diagonal Left-Down check for conflict
    i = row;
    j = col;
    while (i >= 0 && j >= 0) {
      if (board[i][j] == 1) return false;
      i--;
      j--;
    }

    // Diagonal Left-Up check for conflict
    i = row;
    j = col;
    while (i >= 0 && j < board[i].length) {
      if (board[i][j] == 1) return false;
      i--;
      j++;
    }

    // Diagonal Right Down check for conflict
    i = row;
    j = col;
    while (i < board.length && j >= 0) {
      if (board[i][j] == 1) return false;
      i++;
      j--;
    }
    return true;
  }

  static void print(int board[][], int rows, int col) {
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < col; j++) {
        System.out.print(board[i][j] + " ");
      }
      System.out.println();
    }
  }

  public static void main(String args[]) {
    Scanner sc = new Scanner(System.in);
    System.out.println("Enter rows x columns");
    int rows = sc.nextInt();
    int col = sc.nextInt();
    if (rows != col) {
      System.out.println("Invalid Configuration");
      return;
    }
    int board[][] = new int[rows][col];
    System.out.println(Solve(board, 0));
    print(board, rows, col);
  }
}

File: Recursion and Backtracking/Backtracking/NQueens.js
function nqueens(board, col, n) {
	if(col >= n) console.log(board);

	for(let row = 0; row < n; row++) {
		if (isSafe(board, row, col, n)) {
			placeQueen(board, row, col);
			nqueens(board, col + 1, n)
			removeQueen(board, row, col)
		}
	}
}

function isSafe(board, row, col, n) {
	// vertical up check
	let i = row;
	while (i >= 0) {
		if(board[i][col] === 'Q') return false;
		i--;
	}
	
	i = row;
	// vertical down
	while (i < n) {
		if(board[i][col] === 'Q') return false;
		i++;
	}

	i = col;
	// horizontal left
	while (i >= 0) {
		if(board[row][i] === 'Q') return false;
		i--;
	}

	i = col;
	// horizontal right check
	while (i < n) {
		if(board[row][i] === 'Q') return false;
		i++;
	}

	// diagonal left-up
	let r = row;
	let c = col;
	while (r >= 0 && c >= 0) {
		if (board[r][c] === 'Q') return false;
		r--;
		c--;
	} 

	// diagonal right-down
	r = row;
	c = col;

	while (r < n && c < n) {
		if(board[r][c] === 'Q') return false;
		r++;
		c++;
	}

	// diagonal right-up
	r = row;
	c = col;

	while (r >= 0 && c < n) {
		if(board[r][c] === 'Q') return false
		r--;
		c++;
	}

	// diagonal left-down
	r = row;
	c = col;

	while (r < n && c >= 0) {
		if(board[r][c] === 'Q') return false;
		r++;
		c--;
	}
	return true;
}

function placeQueen(board, row, col) {
	board[row][col] = 'Q'
}

function removeQueen(board, row, col) {
	board[row][col] = '-';
}


const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
})

readline.question('Enter Chess size(n) ', n => {
	let board = [];
	let row = [];
	for (let i = 0; i < n; i++) {
		row.push('-');
	}
	for (let i = 0; i < n; i++) {
		board[i] = [...row];
	}
        nqueens(board, 0, Number(n));
//	for(let i = 0; i < n; i++) {
//		for(let j = 0; j < n; j++) {
//			process.stdout.write(board[i][j]+" ");
//		}
//		console.log();
//	}
        readline.close();
});

